test/org/javarosa/xpath/expr/XPathPathExprCurrentTest.java

```diff
@@ -0,0 +1,100 @@
+import org.junit.Before;
+import org.junit.Test;
+
+import static junit.framework.TestCase.assertEquals;
+import static org.javarosa.test.utils.ResourcePathHelper.r;
+
+public class XPathPathExprCurrentTest {
+    private FormDef formDef;
+    private FormEntryController formEntryController;
+
+    @Before
+    public void setUp() {
+        FormParseInit fpi = new FormParseInit(r("relative-current-ref.xml"));
+        formDef = fpi.getFormDef();
+        formDef.initialize(true, new InstanceInitializationFactory());
+        FormEntryModel formEntryModel = new FormEntryModel(formDef);
+        formEntryController = new FormEntryController(formEntryModel);
+    }
+
+    /**
+     * current() in a calculate should refer to the node it is in (in this case, /data/my_group/name_relative).
+     * This means that to refer to a sibling node, the path should be current()/../<name of sibling node>. This is
+     * verified by changing the value of the node that the calculate is supposed to refer to
+     * (/data/my_group/name) and seeing that the dependent calculate is updated accordingly.
+     */
+    @Test
+    public void current_in_calculate_should_refer_to_node() {
+        formEntryController.jumpToFirstQuestionWithName("name");
+        StringData nameValue = new StringData("Bob");
+        formEntryController.answerQuestion(nameValue, true);
+
+        StringData relativeName = (StringData) formDef.getFirstDescendantWithName("name").getValue();
+        assertEquals(nameValue.getValue(), relativeName.getValue());
+    }
+
+    /**
+     * current() in a choice filter should refer to the select node the choice filter is called from, NOT the expression
+     * it is in. See https://developer.mozilla.org/en-US/docs/Web/XPath/Functions/current -- this is the difference
+     * between current() and .
+     *
+     * The behavior of current() in a choice filter is verified by selecting a value for a first, static select and then
+     * using that value to filter a second, dynamic select.
+     */
+    @Test
+    public void current_in_choice_filter_should_refer_to_node() {
+        formEntryController.jumpToFirstQuestionWithName("fruit");
+        Selection fruitSelection = new Selection(1);
+        fruitSelection.attachChoice(formEntryController.getModel().getQuestionPrompt().getQuestion());
+        formEntryController.answerQuestion(new SelectOneData(fruitSelection), true);
+
+        SelectOneData fruitSelectionValue = (SelectOneData) formDef.getFirstDescendantWithName("fruit").getValue();
+        assertEquals("blueberry", fruitSelectionValue.getDisplayText());
+
+        // === Variety question ===
+        formEntryController.stepToNextEvent();
+        // Collect calls getAnswerValue to populate dynamic choices
+        FormEntryPrompt varietyPrompt = formEntryController.getModel().getQuestionPrompt();
+        varietyPrompt.getAnswerValue();
+        ItemsetBinding dynamicChoices = varietyPrompt.getQuestion().getDynamicChoices();
```

Comment:
`List<SelectChoice> choices = getChoicesOf("/data/variety");`

And then `getChoicesOf()` takes care of calling whatever method needs to be called.

In Briefcase, I managed to [populate dynamic choices](https://github.com/ggalmazor/briefcase/commit/00df91b05a8de61355daeca1d375d23e78437c07#diff-5a1a0d134a81a1a17735fb53c0b5c563R120) by doing this: 
```java
formDef.initialize(false, new InstanceInitializationFactory());
formDef.populateDynamicChoices(control.getDynamicChoices(), (TreeReference) control.getBind().getReference());
```

I don't know if it could be applied here.

Reply:
The form def initialization is in the `before` block.

`populateDynamicChoices` is called by `getAnswerValue()`. I did it this way to most closely match what Collect does (it doesn't call `populateDynamicChoices` directly).

Reply:
So, the main advantage of using the `populateDynamicChoices` versus the `getAnswerValue()` is that it doesn't depend on where the current form index is located, making it less prone to error and easier to understand.